### ISSUE TITLE
gevent's subprocess stdout as pipe processing blocks the event loop

### ISSUE THREAD

**Issue by tzickel** [gevent/gevent#1336]
* gevent version: 1.3.7
* Python version: 2.7 latest on ubuntu 16.04 and 3.7.1 from homebrew on mac
* Operating System: Written above on x64

### Description:

I am trying to log the stdout / stderr of a subprocess via a greenlet loop (one for each), when there is heavy writing i/o on the child process, gevent stops processing other greenlets on the event loop, and thus they are starved until the i/o pressure stops.

I expected that the event loop would not get stuck...

### What I've run:

```python
from gevent import sleep, spawn
import gevent.subprocess

def line(out):
    for line in iter(out.readline, b''):
        pass

proc = gevent.subprocess.Popen('yes hi', shell=True, stdout=gevent.subprocess.PIPE)
spawn(line, proc.stdout)
sleep(1)
print('I never get here')
proc.kill()
```

**Comment by jamadden** [comment]
The file descriptor you're looping on is effectively *always* ready to read, so you never yield to the event loop (there's never any need to go to a `select()` to wait for data to become ready, the answer is always "yes"). (The same can be seen reading from a regular file, which is also always ready to read; local sockets can demonstrate the same effect too) You are essentially CPU bound, just as if the loop was `while True: pass`.  

Loops that are CPU bound need to yield to the event loop periodically (a call to `sleep()` with no parameters) to allow resource sharing. How frequently that can or should be done is application specific; some might want to do it on every iteration of the loop, some might want to do it every *x* iterations or after *x* amount of time has passed.

**Comment by arcivanov** [comment]
Just curious, did you monkeypatch @tzickel?

**Comment by arcivanov** [comment]
@jamadden I disagree. I have had several servers working with pipes and FIFOs assigned to subprocess' FDs and those FDs are not always ready to read if they are properly put in a non-blocking mode via fcntl.

**Comment by arcivanov** [comment]
@tzickel if the processes produces such a volume of IO that `select` never blocks, there's always going to be data to read in `line` func. Putting `sleep 1` into the subprocess causes `select` to block which in turn causes greenlet switch and everything `works`.

That said, I was under the impression that occasional loop yielding in tight select loops to make sure there was no starvation was already added to gevent a few versions ago. 

Meanwhile changing to:

```
    for line in iter(out.readline, b''):
        sleep(0)
```

will solve your problem.

**Comment by arcivanov** [comment]
Yep, got it, thanks. I wonder if there is an efficient way to yield out of read/write based on the number of pending greenlets. Maybe a per-greenlet counter that checks every thousand IOPS if it show be polite and yield. Checking/incrementing a single int shouldn't be that much of an overhead per IOP and every 1/1000th deeper check if there are other pending spawns shouldn't drop down the average performance noticeably.

**Comment by AndCycle** [comment]
got some probably related issue with gevent 1.4.0 under python 3.7.2,

as I use patched `subprocess.run` to execute `ykchalresp.exe` 
which is a CLI use for YubiKey which is a USB device for challenge response,

the code is simple as
```
from gevent import monkey; monkey.patch_all()
import subprocess
result = subprocess.run(['ykchalresp.exe', '-2', '-t', '-6'],  stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```

the command result in a timeout error report by `ykchalresp.exe` with python 3.7.2
```
>> result.stderr
b'Yubikey core error: timeout\r\n'
```


the weird thing is this works fine with python 3.6.8, no idea about where to start,
just left a note here as I haven't figure it out is this a separate issue or what.

to replicate the issue require a YubiKey which I think not much people have one.
